gapic-common/lib/gapic/grpc/service_stub/channel_pool.rb

```diff
@@ -0,0 +1,113 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "grpc"
+require "googleauth"
+require "gapic/config"
+
+
+module Gapic
+  class ServiceStub
+    ##
+    # Gapic gRPC ServiceStub ChannelPool
+    #
+    # This class wraps multiple channels for sending RPCs.
+    #
+    class ChannelPool
+      ##
+      # Initialize an instance of ServiceStub::ChannelPool
+      #
+      def initialize grpc_stub_class, endpoint:, credentials:, channel_args: nil, interceptors: nil, config: nil
+        if credentials.is_a? ::GRPC::Core::Channel
+          raise ArgumentError, "Can't create a channel pool with GRPC::Core::Channel as credentials"
+        end
+
+        @grpc_stub_class = grpc_stub_class
+        @endpoint = endpoint
+        @credentials = credentials
+        @channel_args = channel_args
+        @interceptors = interceptors
+        @config = config || Configuration.new
+
+        yield @config if block_given?
+
+        @channels = (1..@config.channel_count).map { create_channel }
+      end
+
+      ##
+      # Creates a new channel.
+      def create_channel
+        Channel.new @grpc_stub_class, endpoint: @endpoint, credentials: @credentials, channel_args: @channel_args,
+                    interceptors: @interceptors, on_channel_create: @config.on_channel_create
+      end
+
+      ##
+      # Invoke the specified RPC call.
+      #
+      # @param method_name [Symbol] The RPC method name.
+      # @param request [Object] The request object.
+      # @param options [Gapic::CallOptions, Hash] The options for making the RPC call. A Hash can be provided to
+      #   customize the options object, using keys that match the arguments for {Gapic::CallOptions.new}. This object
+      #   should only be used once.
+      #
+      # @yield [response, operation] Access the response along with the RPC operation.
+      # @yieldparam response [Object] The response object.
+      # @yieldparam operation [::GRPC::ActiveCall::Operation] The RPC operation for the response.
+      #
+      # @return [Object] The response object.
+      #
+      def call_rpc method_name, request, options: nil, &block
+        case @config.channel_selection.downcase.to_sym
+        when :least_loaded
```

Comment:
Is there a default behavior? What happens if `channel_selection` is any other value (or if it's unset or nil)?

Reply:
When using `config_attr`, the default value is returned even when the user explicitly sets the value to nil.

Reply:
If we never expect it to be anything other than `:least_loaded`, we can get rid of the case statement check (and it makes me wonder if we even need that config field). If we think there might be other values in the future, we should provide a fallback `else` clause to handle those: either providing a default value, or raising an exception (or at least emitting a warning) to note something unexpected is going on. Just defensive programming: currently it does nothing silently if it ever happens, and that becomes very difficult to diagnose.

Reply:
We can remove the case block for now, as `channel_selection` has a single valid value. But lets keep the `channel_selection` config field, future requirements might need various algorithms like `round_robin` etc for channel selection. 
Incase of invalid values, we can emit a warning and resort to least loaded channel selection.